### PR TITLE
Improve message Access Denied: insufficient view owner privileges

### PR DESCRIPTION
**Pull request by zaz968m**
If a user A creates a view and user B wants to access the view and then user A loses access to the original data then B should not be able to access data in the view.
If B gets and access denied, they are told whether it's because they don't have access to the view or if the view creator doesn't have access to the underlying data.
This PR is only about enhancing information in the error message.

### CODE REVIEW

File: presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java [hunk 1]
```diff
@@ -321,6 +326,11 @@ public static void denyCreateViewWithSelect(String sourceName, ConnectorIdentity
         throw new AccessDeniedException(format("View owner '%s' cannot create view that selects from %s%s", identity.getUser(), sourceName, formatExtraInfo(extraInfo)));
     }
 
+    public static void denySelectFromColumns(AccessDeniedException e)
+    {
+        throw new AccessDeniedException("View owner does not have sufficient privileges", e);
```
Comment (sopel39) [inline presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java:331]: this seems generic and not specific to columns only. All column names available (accessible) to query user?

Comment (zaz968m) [reply presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java:331]: Thank you, renamed.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java [hunk 2]
```diff
@@ -14,13 +14,15 @@
 package io.prestosql.security;
```
[the remark this answers is not in the document]
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:14]: done

File: presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java [hunk 3]
```diff
@@ -28,7 +28,12 @@
 {
     public AccessDeniedException(String message)
     {
-        super(PERMISSION_DENIED, "Access Denied: " + message);
+        this(message, null);
+    }
+
+    public AccessDeniedException(String message, Throwable throwable)
```
[the remark this answers is not in the document]
Comment (zaz968m) [reply presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java:34]: done

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java [hunk 4]
```diff
@@ -295,6 +312,9 @@ protected void assertAccessDenied(
             }
             catch (RuntimeException e) {
                 assertExceptionMessage(sql, e, ".*Access Denied: " + exceptionsMessageRegExp);
+                if (exceptionsCauseMessageRegExp != null) {
+                    assertCauseExceptionMessage(sql, e, ".*Access Denied: " + exceptionsCauseMessageRegExp);
```
Comment (electrum) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:316]: Use AssertJ instead of something custom

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java [hunk 5]
```diff
@@ -42,7 +44,12 @@ public void checkCanSelectFromColumns(SecurityContext context, QualifiedObjectNa
         // In SQL, views are special in that they execute with permissions of the owner.
         // This means that the owner of the view is effectively granting permissions to the user running the query,
         // and thus must have the equivalent of the SQL standard "GRANT ... WITH GRANT OPTION".
-        delegate.checkCanCreateViewWithSelectFromColumns(context, tableName, columnNames);
+        try {
+            delegate.checkCanCreateViewWithSelectFromColumns(context, tableName, columnNames);
+        }
+        catch (AccessDeniedException e) {
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:50]: We should fix the callers instead of catching and rethrowing here

**Review by electrum — CHANGES_REQUESTED** [review]

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java (hunk 4 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:316]: Sure! I thought it is not used in this class for some reason.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 5 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:50]: @electrum In this case we clearly state that it's OWNER(delegate) who lacks privileges and it's possible to find why he lacks these privileges (as stacktrace is preserved). 

Otherwise functionality of AccessControlManager#checkCanCreateViewWithSelectFromColumns has to be duplicated (at least partially, e.g. AccessControlManager#checkCanAccessCatalog) to state that it is not current user, but view owner lacks privileges (at least to access catalog). Such duplication will produce a lot of code which does absolutely the same stuff.

Comment (electrum) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:50]: Maybe I'm not understanding the goal if this PR. Can you update the description to explain what "improve" means?

File: presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java [hunk 6]
```diff
@@ -321,6 +326,11 @@ public static void denyCreateViewWithSelect(String sourceName, ConnectorIdentity
         throw new AccessDeniedException(format("View owner '%s' cannot create view that selects from %s%s", identity.getUser(), sourceName, formatExtraInfo(extraInfo)));
     }
 
+    public static void denyAccessToViewBecauseOwnerLacksPrivileges(AccessDeniedException e)
```
Comment (electrum) [inline presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java:329]: I don't think this public method makes sense here, as it's only usage is inside `ViewAccessControl`. The other methods are intended for use by any access control implementation. Instead, we can do this inline in `ViewAccessControl`.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 5 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:50]: done.

File: presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java (hunk 6 above)
Comment (zaz968m) [reply presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java]: done.

File: presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java (hunk 1 above)
Comment (zaz968m) [reply presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java]: I'm not sure about column names. Visibility of column names are out of scope of this pr.

**Comment by zaz968m** [comment]
@electrum @kokosing all comments are addressed.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java [hunk 7]
```diff
@@ -74,4 +78,14 @@ public void checkCanGrantExecuteFunctionPrivilege(SecurityContext context, Strin
...
         return delegate.getColumnMasks(context, tableName, columnName, type);
     }
+
+    private void executeAsViewOwner(Consumer<AccessControl> consumer)
+    {
+        try {
+            consumer.accept(delegate);
+        }
+        catch (AccessDeniedException e) {
+            throw new PrestoException(PERMISSION_DENIED, "Access Denied: View owner does not have sufficient privileges", e);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:88]: This:

```suggestion
            throw new PrestoException(PERMISSION_DENIED, "Access Denied: View owner does not have sufficient privileges: " + e.getMessage(), e);
```

would be more useful & more actionable.

**Comment by findepi** [comment]
I am concerned that after the change, it's harder to know which permissions need to be changed in case of "Access Denied".
Imagine a situation when a BI user gets this error for a query and they turn to their sys admin / data engineer for help. Will sys admin know what permissions would need to be change for the user?
cc @willmostly

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 7 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:88]: @findepi Could you please note in what case it will be more useful and what additional value it brings?
I think it is not and as the same information is stored in the rethrown exception e and stores in stacktrace.

**Comment by zaz968m** [comment]
@findepi when a BI user gets this error for a query and they turn to their sys admin / data engineer for help. Sys admin will see the very same information they see right now, but in `caused by` part of the exception.

**Comment by findepi** [comment]
> @findepi when a BI user gets this error for a query and they turn to their sys admin / data engineer for help. Sys admin will see the very same information they see right now, but in `caused by` part of the exception.

I would not assume BI tools are nice enough to retain full stacktrace, unless someone with more hands on experience tells me I should not be concerned.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 7 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java]: updated.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java [hunk 8]
```diff
@@ -74,4 +78,14 @@ public void checkCanGrantExecuteFunctionPrivilege(SecurityContext context, Strin
     {
         return delegate.getColumnMasks(context, tableName, columnName, type);
     }
+
+    private void executeAsViewOwner(Consumer<AccessControl> consumer)
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:82]: This could take `Runnable` since the caller already has the delegate. Then this method can be static.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java [hunk 9]
```diff
@@ -74,4 +78,14 @@ public void checkCanGrantExecuteFunctionPrivilege(SecurityContext context, Strin
...
+    {
+        try {
+            consumer.accept(delegate);
+        }
+        catch (AccessDeniedException e) {
+            throw new PrestoException(PERMISSION_DENIED, "Access Denied: View owner does not have sufficient privileges: " + e.getMessage(), e);
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:88]: This will result in a repeated "Access Denied" in the message:

> Access Denied: View owner does not have sufficient privileges: Access Denied: View owner 'xxx' cannot create view that selects from yyy

We should strip that off, since we know the original message always has it.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java [hunk 10]
```diff
@@ -293,10 +293,26 @@ protected void assertAccessDenied(@Language("SQL") String sql, @Language("RegExp
         assertAccessDenied(getSession(), sql, exceptionsMessageRegExp, deniedPrivileges);
     }
 
+    protected void assertAccessDenied(@Language("SQL") String sql, @Language("RegExp") String exceptionsMessageRegExp, @Language("RegExp") String exceptionsCauseMessageRegExp,
+            TestingPrivilege... deniedPrivileges)
```
Comment (electrum) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:297]: Either wrap all arguments or none

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java [hunk 11]
```diff
@@ -306,7 +322,11 @@ protected void assertAccessDenied(
                 fail("Expected " + AccessDeniedException.class.getSimpleName());
             }
             catch (RuntimeException e) {
-                assertExceptionMessage(sql, e, ".*Access Denied: " + exceptionsMessageRegExp);
+                String query = format("Query: %s", sql);
```
Comment (electrum) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:325]: No need to use string format for simple concatenation
```java
String query = "Query: " + sql;

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 8 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:82]: I considered Runnable as well, however I did not find good method name.

`executeAsViewOwner(() -> delegate.checkCanCreateViewWithSelectFromColumns(context, tableName, columnNames));`
does not sound good, because it does not execute something as owner it just wraps it to owner error. 

`wrapWithOwnerException(() -> delegate.checkCanCreateViewWithSelectFromColumns(context, tableName, columnNames));` is not good because you can write whatever you want within runnable and it will be just wraped with exception. No relation between delegate(view owner) and exception at all.

`executeAsViewOwner(owner -> owner.checkCanCreateViewWithSelectFromColumns(context, tableName, columnNames));` I feel that this code somehow obliges me to utilize owner parameter. Once I use it - I have an exception related to this parameter.
Despite this version is a little bit longer I prefer it, unless you have better suggestion.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java (hunk 10 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:297]: done

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 9 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:88]: Done

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java (hunk 11 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:325]: done

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java [hunk 12]
```diff
@@ -74,4 +78,14 @@ public void checkCanGrantExecuteFunctionPrivilege(SecurityContext context, Strin
...
         return delegate.getColumnMasks(context, tableName, columnName, type);
     }
+
+    private void executeAsViewOwner(Consumer<AccessControl> consumer)
+    {
+        try {
+            consumer.accept(delegate);
+        }
+        catch (AccessDeniedException e) {
+            throw new PrestoException(PERMISSION_DENIED, "View owner does not have sufficient privileges: " + e.getMessage(), e);
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:88]: Please throw `io.prestosql.spi.security.AccessDeniedException`

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 8 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:82]: > I considered Runnable as well, however I did not find good method name.


Let's call it `wrapAccessDeniedException`

**Review by kokosing — DISMISSED** [review]
% comments

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 12 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:88]: I think that AccessDeniedException 
will be inconsistent with this conversation: https://github.com/prestosql/presto/pull/5086#discussion_r489786857
and relates a little bit to https://github.com/prestosql/presto/pull/5086#discussion_r486670501

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 8 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:82]: I still don't like it, but 2:1. Renamed.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 12 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:88]: Thanks

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java [hunk 13]
```diff
@@ -74,4 +78,14 @@ public void checkCanGrantExecuteFunctionPrivilege(SecurityContext context, Strin
     {
         return delegate.getColumnMasks(context, tableName, columnName, type);
     }
+
+    private void wrapAccessDeniedException(Runnable runnable)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:82]: static

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 8 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java]: @electrum why do you prefer static over non static methods?

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 13 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java]: done

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 8 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java]: static method is just a function while method is a method ;) 

 - Function does not requires polymorphism and it is much easier for compiler and runtime.
 - It is much easier to maintain that, one can easily extract this function out of this class and reuse it somewhere.
 - It is more readable, reader knows that this function won't modify or even access the object state.

**Comment by zaz968m** [comment]
@kokosing @electrum all comments are addressed

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java [hunk 14]
```diff
@@ -294,9 +293,28 @@ protected void assertAccessDenied(@Language("SQL") String sql, @Language("RegExp
...
+            TestingPrivilege... deniedPrivileges)
+    {
+        assertAccessDenied(getSession(), sql, ".*Access Denied: " + exceptionsMessageRegExp, null, deniedPrivileges);
+    }
+
+    private void assertAccessDenied(
             Session session,
             @Language("SQL") String sql,
             @Language("RegExp") String exceptionsMessageRegExp,
+            @Language("RegExp") String exceptionsCauseMessageRegExp,
```
Comment (findepi) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:317]: i'd suggest decoupling running query with verifying the exception using 
`org.assertj.core.api.Assertions#assertThatThrownBy(org.assertj.core.api.ThrowableAssert.ThrowingCallable)`

a helper like `io.prestosql.testing.assertions.PrestoExceptionAssert#assertPrestoExceptionThrownBy` can be useful

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java [hunk 15]
```diff
@@ -1062,6 +1062,7 @@ public void testViewColumnAccessControl()
         // verify selecting from a view over a table requires the view owner to have special view creation privileges for the table
         assertAccessDenied(
                 "SELECT * FROM " + columnAccessViewName,
+                "View owner does not have sufficient privileges",
```
Comment (findepi) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1065]: can we match the full exception message here, not just a aubstring?
(the actual message has `: ...` after "privileges")

this would also make sure @electrum is happy about "Access Denied" being or not being repeated within the message

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java [hunk 16]
```diff
@@ -294,9 +293,28 @@ protected void assertAccessDenied(@Language("SQL") String sql, @Language("RegExp
     }
 
     protected void assertAccessDenied(
+            @Language("SQL") String sql,
+            @Language("RegExp") String exceptionsMessageRegExp,
+            @Language("RegExp") String exceptionsCauseMessageRegExp,
+            TestingPrivilege... deniedPrivileges)
+    {
+        assertAccessDenied(getSession(), sql, ".*" + exceptionsMessageRegExp + ": Access Denied: " + exceptionsCauseMessageRegExp, ".*Access Denied: " + exceptionsCauseMessageRegExp, deniedPrivileges);
```
Comment (kokosing) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:301]: Please do not use `.*` at the begin and end of error message regexp. We need be sure if there is nothing extra here.

See: 
```
           throw new PrestoException(PERMISSION_DENIED, "View owner does not have sufficient privileges: " + e.getMessage(), e);
```
There is a suffix added that you didn't verify. Please fix this.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 9 above)
Comment (findepi) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java]: does the exception message contain "Access Denied" i once now?

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java [hunk 17]
```diff
@@ -306,7 +324,11 @@ protected void assertAccessDenied(
                 fail("Expected " + AccessDeniedException.class.getSimpleName());
             }
             catch (RuntimeException e) {
-                assertExceptionMessage(sql, e, ".*Access Denied: " + exceptionsMessageRegExp);
+                String query = "Query: " + sql;
+                assertThat(e).as(query).hasMessageMatching(exceptionsMessageRegExp);
+                if (exceptionsCauseMessageRegExp != null) {
```
Comment (kokosing) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:329]: Actually we do not care about the cause, as this is not printed to the user. Please undo these changes.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java (hunk 16 above)
Comment (kokosing) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:301]: I shouldn't be that strict. Maybe `.*` at the beginning might be needed, but not at the end.

Comment (findepi) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:301]: It's better not to use `.*` at all. If there is a spercific part that's variable and cannot be matched (eg table name), using `\w+` is better.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java (hunk 15 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1065]: actually I already had such version https://github.com/prestosql/presto/commit/d2e7d1a92d21d7927b7fdd4ef90c86aabab39ece
it requires adding `: Access Denied: ` somewhere within the error message which is not intuitive and error prone.
If you prefer this way I'll remove `Access Denied` text concatenation from internals of assertAccessDenied methods and add it to every text message during invocation of these methods. At least it will not be misleading.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 9 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java]: Yes.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java (hunk 16 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:301]: done.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java (hunk 15 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1065]: done

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java (hunk 17 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:329]: done

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java (hunk 14 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:317]: done

**Comment by zaz968m** [comment]
@kokosing @electrum @findepi all comments are addressed

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java [hunk 18]
```diff
@@ -1060,9 +1060,9 @@ public void testViewColumnAccessControl()
                 privilege("orders", CREATE_VIEW_WITH_SELECT_COLUMNS));
 
         // verify selecting from a view over a table requires the view owner to have special view creation privileges for the table
-        assertAccessDenied(
+        assertWrappedAccessDenied(
                 "SELECT * FROM " + columnAccessViewName,
-                "View owner 'test_view_access_owner' cannot create view that selects from .*.orders.*",
+                "View owner does not have sufficient privileges: Access Denied: View owner 'test_view_access_owner' cannot create view that selects from \\w+.\\w+.orders\\w*",
```
Comment (kokosing) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1065]: The full error message is:
```
Access Denied: View owner does not have sufficient privileges: Access Denied: View owner 'test_view_access_owner' cannot create view that selects from \\w+.\\w+.orders\
```

There are two issues here.
 - `Access Denied: ` is unnecessarly doubled
 - error message is long and detailed (but on the hand it could be useful to understand why just happened). How about raising only `Access Denied: View owner does not have sufficient privileges` without the second part? @electrum @findepi thoughts?

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java [hunk 19]
```diff
@@ -1060,9 +1060,9 @@ public void testViewColumnAccessControl()
                 privilege("orders", CREATE_VIEW_WITH_SELECT_COLUMNS));
 
         // verify selecting from a view over a table requires the view owner to have special view creation privileges for the table
-        assertAccessDenied(
+        assertWrappedAccessDenied(
```
Comment (kokosing) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1063]: Why `assertAccessDenied` is not enough? Why do you care it is wrapped? To me it is implementation detail on which test should not rely on.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java [hunk 20]
```diff
@@ -299,19 +301,19 @@ protected void assertAccessDenied(
...
+    }
+
+    protected void assertWrappedAccessDenied(@Language("SQL") String sql, @Language("RegExp") String exceptionsMessageRegExp, TestingPrivilege... deniedPrivileges)
+    {
+        assertException(getSession(), sql, exceptionsMessageRegExp, deniedPrivileges);
+    }
+
+    private void assertException(Session session, @Language("SQL") String sql, @Language("RegExp") String exceptionsMessageRegExp, TestingPrivilege[] deniedPrivileges)
+    {
+        assertThatThrownBy(() -> executeExclusively(session, sql, deniedPrivileges))
```
Comment (kokosing) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:314]: This refactor is nice, but it has nothing to do with the commit. Please extract it to separate commit.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java (hunk 18 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1065]: assertWrappedAccessDenied checks the full message ("View owner does not have sufficient privileges: Access Denied: View owner 'test_view_access_owner' cannot create view that selects from \\w+.\\w+.orders\\w*")
1. already resolved here https://github.com/prestosql/presto/pull/5086#discussion_r489786857

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java (hunk 19 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1063]: `assertWrappedAccessDenied` checks the full message.
`assertAccessDenied` cannot be reused because it adds `".*Access Denied: "` at the beginning

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java (hunk 20 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:314]: done

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java (hunk 18 above)
Comment (kokosing) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1065]: 1. Oh... I get it now.

I am not sure if it is nice looking error message:
`View owner does not have sufficient privileges: Access Denied: View owner 'test_view_access_owner' cannot create view `
vs
`
Access Denied: View owner does not have sufficient privileges: View owner 'test_view_access_owner' cannot create view`
vs
`Access Denied: View owner does not have sufficient privileges`

I vote for the last one. If one wants more details then in CLI can use `--debug` in web ui can see the stack trace in web ui.

Comment (findepi) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1065]: The example is... well, an example, right?
The second part can be many different things, so it may be less or more "boring".

My point is -- I don't want to be faced with "Access Denied: View owner does not have sufficient privileges" (period) error message and be asked by the user "ok, so what do i need to change?". It user's query is a big one, it may take a while until you learn which view, which view owner, and what permission they are missing.

**Comment by zaz968m** [comment]
The very same pull req to the forked master: 
https://github.com/zaz968m/presto/pull/1
Some checks were not successful
37 successful and 2 cancelled checks

**Comment by zaz968m** [comment]
Unfortunately I don't have permissions to restart jobs manually. I can immitate it only with force push. So I did one.
Before last force push there was the very same diff https://github.com/prestosql/presto/commit/679b861429a37e333d69127ac209fb5628a9d734
with only one failing test https://github.com/prestosql/presto/runs/1134114335
Running command `./mvnw test -B -fae -Dair.check.skip-all -pl presto-raptor-legacy` for that test locally I have 
```
[INFO] Results:
[INFO] 
[WARNING] Tests run: 641, Failures: 0, Errors: 0, Skipped: 4
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
```

**Comment by findepi** [comment]
restarted the checks

**Comment by zaz968m** [comment]
@electrum @findepi @kokosing are there any open questions left that require actions from my side?

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java (hunk 18 above)
Comment (kokosing) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1065]: Ok, let it be `Access Denied: View owner does not have sufficient privileges: View owner 'test_view_access_owner' cannot create view`.

**Comment by kokosing** [comment]
> are there any open questions left that require actions from my side?

https://github.com/prestosql/presto/pull/5086/commits/9e03d752aea95aff37d303b4c40a981e03649ad6#r490878495

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java (hunk 19 above)
Comment (kokosing) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java:1063]: > assertAccessDenied cannot be reused because it adds ".*Access Denied: " at the beginning

And that is the point. `Access Denied: ` should be placed at the beginning

Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java]: done.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java (hunk 18 above)
Comment (zaz968m) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java]: done.

**Review by kokosing — APPROVED** [review]

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java [hunk 21]
```diff
@@ -74,4 +77,16 @@ public void checkCanGrantExecuteFunctionPrivilege(SecurityContext context, Strin
...
     }
+
+    private static void wrapAccessDeniedException(Runnable runnable)
+    {
+        try {
+            runnable.run();
+        }
+        catch (AccessDeniedException e) {
+            String prefix = AccessDeniedException.PREFIX;
+            String msg = e.getMessage().substring(prefix.length());
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:88]: Please also add `verify(e.getMessage().startsWith(prefix))`

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java [hunk 22]
```diff
@@ -74,4 +77,16 @@ public void checkCanGrantExecuteFunctionPrivilege(SecurityContext context, Strin
...
         return delegate.getColumnMasks(context, tableName, columnName, type);
     }
+
+    private static void wrapAccessDeniedException(Runnable runnable)
+    {
+        try {
+            runnable.run();
+        }
+        catch (AccessDeniedException e) {
+            String prefix = AccessDeniedException.PREFIX;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:87]: nit: I would inline it (up to you, no change request)

**Review by kokosing — APPROVED** [review]

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 22 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:87]: I've considered this as well, but I see two points against it:
1. it will be used in several places, so no 'link' between these usages.
2. `AccessDeniedException.PREFIX` is too long. Once static imported - it's not explicit what does `PREFIX` mean.
Will left it as is.

File: presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java (hunk 21 above)
Comment (zaz968m) [reply presto-main/src/main/java/io/prestosql/security/ViewAccessControl.java:88]: done

**Comment by kokosing** [comment]
@zaz968m There are legitimate test failures:
```
2020-09-28T12:41:41.5110410Z tests               | 2020-09-28 18:26:41 INFO: FAILURE     /    io.prestosql.tests.hive.TestSqlStandardAccessControlChecks.testAccessControlCreateView (Groups: authorization, profile_specific_tests) took 0.5 seconds
2020-09-28T12:41:41.5140795Z tests               | 2020-09-28 18:26:41 SEVERE: Failure cause:
2020-09-28T12:41:41.5142271Z tests               | java.lang.AssertionError: Query failed with unexpected error message: 'java.sql.SQLException: Query failed (#20200928_124141_00912_zcajp): Access Denied: View owner does not have sufficient privileges: View owner 'bob' cannot create view that selects from default.alice_owned_table' 
2020-09-28T12:41:41.5143874Z tests               |  Expected error message to contain 'Access Denied: View owner 'bob' cannot create view that selects from 
```

**Comment by kokosing** [comment]
Failed with: https://github.com/prestosql/presto/issues/5354 and https://github.com/prestosql/presto/issues/4973

**Comment by kokosing** [comment]
Merged, thanks!
